TypeError: 'type' object is not subscriptable

The recent release of 0.4.7 broke my build in Python 3.8:

```
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/macrokit/__init__.py:8: in <module>
    from macrokit.expression import (
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/macrokit/expression.py:224: in <module>
    class Expr:
/usr/share/miniconda/envs/test/lib/python3.8/site-packages/macrokit/expression.py:411: in Expr
    cls, fn: "_Expr", args: tuple[_Expr, ...], kwargs: dict[str, _Expr]
E   TypeError: 'type' object is not subscriptable
```

You're indexing a type:
https://github.com/hanjinliu/macro-kit/blob/b0764e16da0cdf626e8c129fcf05891681f6d5f0/macrokit/expression.py#L411

But also supporting Python 3.8:
https://github.com/hanjinliu/macro-kit/blob/b0764e16da0cdf626e8c129fcf05891681f6d5f0/pyproject.toml#L11

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -38,4 +38,6 @@ jobs:
       - name: Test
         uses: aganders3/headless-gui@v1
         with:
-          run: hatch -v run test:run
+          run: |
+            python -m pip install .[testing] --upgrade hatch
+            hatch -v run test:run

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -1,0 +1,41 @@
+name: tests
+
+on:
+  push:
+    branches:
+      - main
+    tags:
+      - "v*" # Push events to matching v*, i.e. v1.0, v20.15.10
+  pull_request:
+    branches:
+      - main
+  workflow_dispatch:
+
+jobs:
+  test:
+    name: ${{ matrix.platform }} (${{ matrix.python-version }})
+    runs-on: ${{ matrix.platform }}
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        platform: [ubuntu-latest, macos-latest, windows-latest]
+        include:
+          - python-version: "3.10"
+            platform: ubuntu-latest
+
+    steps:
+      - name: Cancel Previous Runs
+        uses: styfle/cancel-workflow-action@0.11.0
+        with:
+          access_token: ${{ github.token }}
+
+      - uses: actions/checkout@v3
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Test
+        uses: aganders3/headless-gui@v1
+        with:
+          run: hatch -v run test:run

--- a/macrokit/expression.py
+++ b/macrokit/expression.py
@@ -431,8 +431,8 @@ class Expr:
         cls,
         obj: "_Expr",
         func: str,
-        args: tuple[_Expr, ...],
-        kwargs: dict[str, _Expr],
+        args: "tuple[_Expr, ...]",
+        kwargs: "dict[str, _Expr]",
     ) -> "Expr":
         """Unsplit ``obj.func(*args, **kwargs)`` to (obj, func, args, kwargs)."""
         fn = cls(Head.getattr, [obj, Symbol(func)])

--- a/macrokit/expression.py
+++ b/macrokit/expression.py
@@ -580,7 +580,7 @@ class Expr:
                 return True
         return False
 
-    def iter_args(self) -> Iterator[Symbol]:
+    def iter_args(self) -> "Iterator[Symbol]":
         """Recursively iterate along all the arguments."""
         for arg in self.args:
             if isinstance(arg, Expr):
@@ -590,7 +590,7 @@ class Expr:
             else:
                 raise RuntimeError(f"{arg} (type {type(arg)})")
 
-    def iter_expr(self) -> Iterator["Expr"]:
+    def iter_expr(self) -> "Iterator[Expr]":
         """
         Recursively iterate over all the nested Expr, until reaching to non-nested Expr.
 

--- a/macrokit/expression.py
+++ b/macrokit/expression.py
@@ -408,7 +408,7 @@ class Expr:
 
     @classmethod
     def unsplit_call(
-        cls, fn: "_Expr", args: tuple[_Expr, ...], kwargs: dict[str, _Expr]
+        cls, fn: "_Expr", args: "tuple[_Expr, ...]", kwargs: "dict[str, _Expr]"
     ) -> "Expr":
         """Unsplit ``func(*args, **kwargs)`` to (func, args, kwargs)."""
         inputs = [fn] + cls._convert_args(args, kwargs)
@@ -468,7 +468,7 @@ class Expr:
         _kwargs = arguments[i:]
 
         # prepare namespaces
-        args_ns: dict[Union[str, _Expr], Any] = ns.copy()
+        args_ns: "dict[Union[str, _Expr], Any]" = ns.copy()
         args_ns[symbol(_tuple)] = _tuple
         kwargs_ns = ns.copy()
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -24,6 +24,13 @@ dependencies = [
     "typing_extensions",
 ]
 
+[project.optional-dependencies]
+testing = [
+    "numpy>=1.21.0",
+    "pandas>=1.3.0",
+    "pytest>=6.2.0",
+]
+
 [project.urls]
 Download = "https://github.com/hanjinliu/macro-kit"
 
@@ -39,3 +46,9 @@ include = [
 include = [
     "/macrokit",
 ]
+
+[tool.hatch.envs.test]
+features = ["testing"]
+
+[tool.hatch.envs.test.scripts]
+run = "pytest -v"

--- a/tests/test_macrokit.py
+++ b/tests/test_macrokit.py
@@ -3,6 +3,7 @@ from macrokit import (
     Expr, Macro, Symbol, parse, register_type, unregister_type, symbol,
     store, store_sequence
 )
+import sys
 import ast
 
 def test_function():
@@ -570,6 +571,7 @@ def test_split_method():
     assert str(expr) == "x.f(1, 2, x=3)"
 
 T = "    "
+@pytest.mark.skipif(sys.version_info < (3, 10), reason="requires python3.10")
 @pytest.mark.parametrize(
     "s",
     [

--- a/tests/test_macrokit.py
+++ b/tests/test_macrokit.py
@@ -295,6 +295,7 @@ def test_binary_op(op: str):
 def test_increment_op(op: str):
     ast.parse(str(parse(f"a {op}= b")))
 
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="ast.unparse not available")
 @pytest.mark.parametrize(
     "s",
     [

